Common/IsolatorLimitResultProvider.cs

```diff
@@ -0,0 +1,110 @@
+    /// </summary>
+    public static class IsolatorLimitResultProvider
+    {
+        /// <summary>
+        /// Provides access to a null implementation of <see cref="IIsolatorLimitResultProvider"/>
+        /// </summary>
+        public static readonly IIsolatorLimitResultProvider Null = new NullIsolatorLimitResultProvider();
+
+        /// <summary>
+        /// Convenience method for invoking a scheduled event's Scan method inside the <see cref="IsolatorLimitResultProvider"/>
+        /// </summary>
+        public static void Consume(
+            this IIsolatorLimitResultProvider isolatorLimitProvider,
+            ScheduledEvent scheduledEvent,
+            DateTime scanTimeUtc
+            )
+        {
+            // perform initial filtering to prevent starting a task when not necessary
+            if (scheduledEvent.NextEventUtcTime > scanTimeUtc)
+            {
+                return;
+            }
+
+            var timeProvider = RealTimeProvider.Instance;
+            isolatorLimitProvider.Consume(timeProvider, scheduledEvent.Name, () => scheduledEvent.Scan(scanTimeUtc));
+        }
+
+        /// <summary>
+        /// Executes the provided code block and while the code block is running, continually consume from
+        /// the limit result provided one token each minute. This function allows the code to run for the
+        /// first full second without requesting additional time from the provider. Following that, every
+        /// minute an additional one minute will be requested from the provider.
+        /// </summary>
+        /// <remarks>
+        /// This method exists to support scheduled events, and as such, intercepts any errors raised via the
+        /// provided code and wraps them in a <see cref="ScheduledEventException"/>. If in the future this is
+        /// usable elsewhere, consider refactoring to handle the errors in a different fashion.
+        /// </remarks>
+        public static void Consume(
+            this IIsolatorLimitResultProvider isolatorLimitProvider,
+            ITimeProvider timeProvider,
+            string name,
+            Action code
+            )
+        {
+            var finished = 0L;
+            Task.Run(async () =>
+            {
+                if (Interlocked.Read(ref finished) != 0L)
+                {
+                    // case when the code block has virtually no code in it and
+                    // was able to complete faster than the task was able to start
+                    return;
+                }
+
+                var next = timeProvider.GetUtcNow().AddMinutes(1);
+                while (Interlocked.Read(ref finished) == 0L)
+                {
+                    if (timeProvider.GetUtcNow() >= next)
+                    {
+                        // each minute request additional time from the isolator
+                        next = next.AddMinutes(1);
```

Comment:
We might want to request time _before_ the minute/time we have is over, to avoid any race conditions with the monitoring thread

Reply:
It sounds like you're referring to potential [nyquist] effects. I'm not sure which monitoring thread you're referring to as there's only a single thread checking times here. The maximum error in this check is based on the sleep interval of this loop, which is currently set to 5ms. So every 5ms we check if a full minute has passed, and if so, we request an extra minute. 

After writing the above, I think there's some misunderstanding here. As soon as the monitoring task starts (_very_ close to when the scheduled event starts), we start checking how long it's been running, every 5ms. When it's been running for a minimum of 1 full minute, we go and request an additional minute. The desired behavior here being that the first minute was free, but the _next_ one, you'll have to pay for. If there are minutes available in the bucket, then we increment to check the following minute, and so on and so forth. When the bucket runs out of minutes _and then_ you request another minute, the `_failed` flag will be flipped and the isolator will terminate the algorithm.